samza-api/src/main/java/org/apache/samza/context/ApplicationContainerContextFactory.java

```diff
@@ -35,14 +36,42 @@
  *
  * @param <T> concrete type of {@link ApplicationContainerContext} created by this factory
  */
+@InterfaceStability.Evolving
 public interface ApplicationContainerContextFactory<T extends ApplicationContainerContext> extends Serializable {
-
   /**
    * Creates an instance of the application-defined {@link ApplicationContainerContext}.
+   * <p>
+   * Applications should implement this to provide a context for container initialization.
    *
+   * @param externalContext external context provided for the application; null if it was not provided
    * @param jobContext framework-provided job context
    * @param containerContext framework-provided container context
    * @return a new instance of the application-defined {@link ApplicationContainerContext}
    */
-  T create(JobContext jobContext, ContainerContext containerContext);
+  default T create(ExternalContext externalContext, JobContext jobContext, ContainerContext containerContext) {
+    return create(jobContext, containerContext);
+  }
+
+  /**
+   * New implementations should not implement this directly. Implement
+   * {@link #create(ExternalContext, JobContext, ContainerContext)} instead.
+   * <p>
+   * This is the same as {@link #create(ExternalContext, JobContext, ContainerContext)}, except it does not provide
+   * access to external context.
+   * <p>
+   * This is being left here for backwards compatibility.
+   *
+   * @param jobContext framework-provided job context
+   * @param containerContext framework-provided container context
+   * @return a new instance of the application-defined {@link ApplicationContainerContext}
+   *
+   * Deprecated: Applications should implement {@link #create(ExternalContext, JobContext, ContainerContext)} directly.
+   * This is being left here for backwards compatibility.
+   */
+  @Deprecated
+  default T create(JobContext jobContext, ContainerContext containerContext) {
+    // adding this here so that new apps do not need to implement this
+    throw new UnsupportedOperationException("Please implement #create(ExternalContext, JobContext, ContainerContext) "
```

Comment:
For existing users already implementing this method, this exception doesn't matter.
For new users, this will only happen if they don't override either create. If the idea is to prevent that, maybe remove ", or at least ..." from the message?

Reply:
I wanted to avoid requiring new users to implement the deprecated version of create, so I needed some default impl for the deprecated version. I was trying to note in the log message that implementing the deprecated version of create was not ideal, but the javadoc should be clear enough, so I'll simplify the log message.